packages/network-controller/src/NetworkController.ts

```diff
@@ -404,6 +430,80 @@ export class NetworkController extends BaseControllerV2<
     }
     return Promise.resolve(true);
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param config - The network configuration.
+   * @param config.rpcUrl - The network configuration RPC URL.
+   * @param config.chainId - The chain ID of the network, as per EIP-155.
+   * @param config.ticker - Currency ticker.
+   * @param config.nickname - Personalized network name.
+   * @param config.rpcPrefs - Personalized preferences.
+   * @returns uuid for the added or updated network configuration
+   */
+  addNetworkConfigurations({
+    rpcUrl,
+    chainId,
+    ticker,
+    nickname,
+    rpcPrefs,
+  }: NetworkConfiguration): string {
+    const networkConfigurations = { ...this.state.networkConfigurations };
+
+    const newNetworkConfiguration: NetworkConfiguration = {
+      rpcUrl,
+      chainId,
+      ticker,
+      nickname,
+      rpcPrefs,
+    };
+
+    // This upsert functionality is equivalent to the preveious impelementation of `addFrequentRpcList` on the PreferencesController
+    // In the long run however we may add additional fields that could mean we can't rely on any individual field - including rpcURL - to be unique
+    // in which case we will need to update this logic to guarantee uniqueness
+    for (const configUUID in networkConfigurations) {
+      if (
+        networkConfigurations[configUUID].rpcUrl.toLowerCase() ===
+        rpcUrl.toLowerCase()
+      ) {
+        // update existing config
+        networkConfigurations[configUUID] = newNetworkConfiguration;
+        this.update((state) => {
+          state.networkConfigurations = networkConfigurations;
+        });
+        return configUUID;
+      }
+    }
```

Comment:
This upsert functionality is equivalent to the previous implementation of `addFrequentRpcList` on the PreferencesController. In the long run however we may add additional fields that could mean we can't rely on any individual field - including `rpcURL` - to be unique in which case we will need to update this logic to guarantee uniqueness.

Reply:
Nit: Should the name reflect this behavior as well? e.g. `upsertNetworkConfiguration` (also dropping the `s` because it only accepts configuration for a single network)